sql/parser/constraint_tree.go

```diff
@@ -0,0 +1,116 @@
+package parser
+
+import (
+	"fmt"
+
+	"github.com/genjidb/genji/document"
+)
+
+type constraintNode struct {
+	frag   document.ValuePathFragment
+	typ    document.ValueType
+	parent *constraintNode
+	sub    []*constraintNode
+}
+
+func createConstraintNode(parent *constraintNode, path document.ValuePath, typ document.ValueType) *constraintNode {
+	node := &constraintNode{
+		frag:   path[0],
+		parent: parent,
+	}
+
+	if len(path) == 1 {
+		node.typ = typ
+		return node
+	}
+
+	// Here we check next path fragment
+	// If next is index, current node is array
+	// Otherwise it is a document
+	if path[1].FieldName != "" {
+		node.typ = document.DocumentValue
+	} else {
+		node.typ = document.ArrayValue
+	}
+
+	node.sub = append(node.sub, createConstraintNode(node, path[1:], typ))
+	return node
+}
+
+func (n *constraintNode) getPath() document.ValuePath {
+	if n.parent == nil {
+		return document.ValuePath{n.frag}
+	}
+
+	return append(n.parent.getPath(), n.frag)
+}
+
+func (n *constraintNode) search(path document.ValuePath) *constraintNode {
+	if n.frag == path[0] {
```

Comment:
I think you should check the non equal `path` first for more readability.
What do you think about that? 
```go
    if n.frag != path[0] {
          return nil
   }
  ....
```